### PR TITLE
feat: integrate tests with backend-sdk-testing

### PR DESCRIPTION
**Pull request by sandrodesouza**
## Summary of change

this PR integrates current tests with the new backend-sdk-testing. Starting with accountlinking and accountlinking-with-session folder.

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here highlighting the necessary changes)

## Checklist for important updates

-   [ ] Changelog has been updated
-   [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `lib/ts/version.ts`
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [ ] Changes to the version if needed
    -   In `package.json`
    -   In `package-lock.json`
    -   In `lib/ts/version.ts`
-   [x] Had run `npm run build-pretty`
-   [x] Had installed and ran the pre-commit hook
-   [ ] If new thirdparty provider is added,
    -   [ ] update switch statement in `recipe/thirdparty/providers/configUtils.ts` file, `createProvider` function.
    -   [ ] add an icon on the user management dashboard.
-   [ ] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [ ] If have added a new web framework, update the `add-ts-no-check.js` file to include that
-   [ ] If added a new recipe / api interface, then make sure that the implementation of it uses NON arrow functions only (like `someFunc: function () {..}`).
-   [ ] If added a new recipe, then make sure to expose it inside the recipe folder present in the root of this repo. We also need to expose its types.

## Remaining TODOs for this PR

### CODE REVIEW

**Review by porcellus — CHANGES_REQUESTED** [review]

File: test/test-server/package.json [hunk 1]
```diff
@@ -0,0 +1,20 @@
+{
+    "name": "test-server",
+    "version": "1.0.0",
+    "description": "",
+    "main": "dist/index.js",
+    "scripts": {
+        "build": "sh ./build.sh",
+        "start": "SUPERTOKENS_ENV=testing TEST_MODE=testing node dist/index.js",
```
[the remark this answers is not in the document]
Comment (sandrodesouza) [reply test/test-server/package.json:8]: nop, I will remove it

File: test/test-server/package.json [hunk 2]
```diff
@@ -0,0 +1,20 @@
+{
+    "name": "test-server",
+    "version": "1.0.0",
+    "description": "",
+    "main": "dist/index.js",
+    "scripts": {
+        "build": "sh ./build.sh",
```
[the remark this answers is not in the document]
Comment (sandrodesouza) [reply test/test-server/package.json:7]: lets go with tsx

File: .circleci/setupAndTestWithFreeCore.sh [hunk 3]
```diff
@@ -55,3 +55,26 @@ if ! [[ -z "${CIRCLE_NODE_TOTAL}" ]]; then
 else
     TEST_MODE=testing SUPERTOKENS_CORE_TAG=$coreTag NODE_PORT=8081 INSTALL_PATH=../supertokens-root npm test
 fi
+
+API_PORT=3030
+ST_CONNECTION_URI=http://localhost:$NODE_PORT
```
[the remark this answers is not in the document]
Comment (sandrodesouza) [reply .circleci/setupAndTestWithFreeCore.sh:60]: no, I removed it

File: test/test-server/src/accountlinking.ts [hunk 4]
```diff
@@ -0,0 +1,161 @@
+import type { Debugger } from "debug";
+import type { Express } from "express";
+import { User } from "../../../lib/build";
+import AccountLinkingRecipe from "../../../lib/build/recipe/accountlinking/recipe";
+import AccountLinking from "../../../recipe/accountlinking";
+import * as supertokens from "../../../lib/build";
+
+export function setupAccountlinkingRoutes(app: Express, log: Debugger) {
```
[the remark this answers is not in the document]
Comment (sandrodesouza) [reply test/test-server/src/accountlinking.ts:8]: yes, make sense

File: test/test-server/src/accountlinking.ts [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    app.post("/test/accountlinking/createprimaryuser", async (req, res, next) => {
```
[the remark this answers is not in the document]
Comment (sandrodesouza) [reply test/test-server/src/accountlinking.ts:9]: yes, I will change it.

File: test/test-server/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+import { getSessionVars, resetSessionVars, setupSessionRoutes } from "./session";
+import { setupSupertokensRoutes } from "./supertokens";
+import * as supertokens from "../../../lib/build";
+import express from "express";
+import nock from "nock";
+import { setupMultiFactorAuthRoutes } from "./multifactorauth";
+import { setupThirdPartyRoutes } from "./thirdparty";
+import { setupTOTPRoutes } from "./totp";
+
+const log = debug("api-mock");
```
[the remark this answers is not in the document]
Comment (sandrodesouza) [reply test/test-server/src/index.ts:47]: ok

File: test/test-server/src/index.ts [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+});
+
+app.use((req, res, next) => {
+    res.status(404).send(`api-mock: route not found ${req.method} ${req.path}`);
+    throw new Error(`api-mock: route not found ${req.method} ${req.path}`);
+});
+
+app.listen(API_PORT, "localhost", () => {
+    log(`api-mock-server started on localhost:${API_PORT}`);
+});
```
[the remark this answers is not in the document]
Comment (sandrodesouza) [reply test/test-server/src/index.ts:399]: ok

File: .circleci/setupAndTestWithFreeCore.sh [hunk 8]
```diff
@@ -55,3 +55,25 @@ if ! [[ -z "${CIRCLE_NODE_TOTAL}" ]]; then
...
+pushd test/test-server
+npm install
+API_PORT=$API_PORT ST_CONNECTION_URI=$ST_CONNECTION_URI npm start &
+popd
+
+# run tests
+cd ../
+git clone git@github.com:supertokens/backend-sdk-testing.git
+cd backend-sdk-testing
+git checkout 641af3fbe8d6b7a064726be82244df3ff4b32bce # checkout latest
```
Comment (porcellus) [inline .circleci/setupAndTestWithFreeCore.sh:72]: Will the FDI based test checkout be done in a separate PR?

**Review by porcellus — CHANGES_REQUESTED** [review]

File: .circleci/setupAndTestWithFreeCore.sh (hunk 8 above)
Comment (sandrodesouza) [reply .circleci/setupAndTestWithFreeCore.sh]: yes

File: test/test-server/src/multifactorauth.ts [hunk 9]
```diff
@@ -1,11 +1,16 @@
...
+import { logger } from "./logger";
 
-export function setupMultiFactorAuthRoutes(app: Express, log: Debugger) {
-    app.post("/test/multifactorauth/multifactorauthclaim.fetchvalue", async (req, res, next) => {
+const namespace = "com.supertokens:node-test-server:multifactorauth";
+const { logDebugMessage } = logger(namespace);
+
+const router = Router()
+    //
+    .post("/multifactorauthclaim.fetchvalue", async (req, res, next) => {
```
[the remark this answers is not in the document]
Comment (sandrodesouza) [reply test/test-server/src/multifactorauth.ts]: ok

File: test/test-server/src/thirdparty.ts [hunk 10]
```diff
@@ -1,12 +1,16 @@
...
+import { logger } from "./logger";
 
-export function setupThirdPartyRoutes(app: Express, log: Debugger) {
-    app.post("/test/thirdparty/manuallycreateorupdateuser", async (req, res, next) => {
+const namespace = "com.supertokens:node-test-server:thirdparty";
+const { logDebugMessage } = logger(namespace);
+
+const router = Router()
+    //
+    .post("/manuallycreateorupdateuser", async (req, res, next) => {
```
[the remark this answers is not in the document]
Comment (sandrodesouza) [reply test/test-server/src/thirdparty.ts]: ok

File: test/test-server/src/index.ts [hunk 11]
```diff
@@ -0,0 +1,401 @@
...
+    nock(url)[method](path).reply(status, body);
+    res.json({ ok: true });
+});
+
+app.get("/test/waitforevent", async (req, res, next) => {
+    try {
+        logDebugMessage("ProcessState:waitForEvent %j", req.query);
+        const instance = ProcessState.getInstance();
+        const eventEnum = req.query.event ? Number(req.query.event) : null;
+        const event = eventEnum ? await instance.waitForEvent(eventEnum) : undefined;
```
Comment (porcellus) [inline test/test-server/src/index.ts:334]: We should make the event param required and throw in that case instead of skipping the entire function and returning immediately. It signifies a test coding error and it could make test pass when they shouldn't.

File: test/test-server/src/index.ts [hunk 12]
```diff
@@ -0,0 +1,401 @@
...
+    const { url, status, body, path, method } = req.body;
+    nock(url)[method](path).reply(status, body);
+    res.json({ ok: true });
+});
+
+app.get("/test/waitforevent", async (req, res, next) => {
+    try {
+        logDebugMessage("ProcessState:waitForEvent %j", req.query);
+        const instance = ProcessState.getInstance();
+        const eventEnum = req.query.event ? Number(req.query.event) : null;
```
Comment (porcellus) [inline test/test-server/src/index.ts:333]: if `req.query.event` is 0 (which it is for `CALLING_SERVICE_IN_VERIFY`) then we'd get the "skip branch" that always returns `undefined`. Please make this an explicit check.

**Review by porcellus — CHANGES_REQUESTED** [review]

File: test/test-server/src/emailpassword.ts [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+                ...("user" in response
+                    ? {
+                          user: response.user.toJson(),
+                      }
+                    : {}),
+                ...("recipeUserId" in response
+                    ? {
+                          recipeUserId: response.recipeUserId.getAsString(),
+                      }
+                    : {}),
```
[the remark this answers is not in the document]
Comment (sandrodesouza) [reply test/test-server/src/emailpassword.ts]: done

File: test/test-server/src/emailverification.ts [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+                      sessionRaw.antiCsrfToken,
+                      sessionRaw.sessionHandle,
+                      sessionRaw.userId,
+                      supertokens.convertToRecipeUserId(sessionRaw.recipeUserId.recipeUserId),
+                      sessionRaw.userDataInAccessToken,
+                      sessionRaw.reqResInfo,
+                      sessionRaw.accessTokenUpdated,
+                      sessionRaw.tenantId
+                  )
+                : sessionRaw;
```
[the remark this answers is not in the document]
Comment (sandrodesouza) [reply test/test-server/src/emailverification.ts]: yes, it is fixed

File: test/test-server/src/session.ts [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+    userIdInCallback = undefined;
+    recipeUserIdInCallback = undefined;
+}
+
+const router = Router()
+    .post("/createnewsessionwithoutrequestresponse", async (req, res, next) => {
+        try {
+            logDebugMessage("Session.createNewSessionWithoutRequestResponse %j", req.body);
+            const recipeUserId = supertokens.convertToRecipeUserId(
+                req.body.recipeUserId.recipeUserId || req.body.recipeUserId
```
[the remark this answers is not in the document]
Comment (sandrodesouza) [reply test/test-server/src/session.ts]: old code, is removed now

File: test/test-server/src/emailverification.ts [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+            const response = await EmailVerificationRecipe.getInstance()?.updateSessionIfRequiredPostEmailVerification({
```
[the remark this answers is not in the document]
Comment (sandrodesouza) [reply test/test-server/src/emailverification.ts]: ok done

File: test/test-server/src/index.ts (hunk 11 above)
Comment (sandrodesouza) [reply test/test-server/src/index.ts]: make sense, done.

File: test/test-server/src/index.ts (hunk 12 above)
Comment (sandrodesouza) [reply test/test-server/src/index.ts]: done

File: test/test-server/src/emailverification.ts [hunk 17]
```diff
@@ -66,26 +66,14 @@ const router = Router()
...
-                      sessionRaw.tenantId
-                  )
-                : sessionRaw;
-            const response = await EmailVerificationRecipe.getInstance()?.updateSessionIfRequiredPostEmailVerification({
+            const session = req.body.session && (await handleSession(req.body.session));
+            let instance = EmailVerificationRecipe.getInstance();
+            if (instance === undefined) {
+                throw new Error("No instance of EmailVerificationRecipe found");
+            }
+            const response = await instance.updateSessionIfRequiredPostEmailVerification({
```
Comment (porcellus) [inline test/test-server/src/emailverification.ts:76]:
```suggestion
            const response = await EmailVerificationRecipe.getInstanceOrThrowError().updateSessionIfRequiredPostEmailVerification({
```

**Review by porcellus — CHANGES_REQUESTED** [review]

File: test/test-server/src/emailverification.ts (hunk 17 above)
Comment (sandrodesouza) [reply test/test-server/src/emailverification.ts]: done
